Matplotlib 3.8 update

Despiix opened the pull request:
### Description of work

#### Summary of work
<!-- Please provide a short, high level description of the work that was done.
--> Update Matplotlib to 3.8.4 from 3.7

<!-- Why has this work been done? If there is no linked issue please provide appropriate context for this work.
#### Purpose of work
This can be removed if a github issue is referenced below
--> 
This pr updates the codebase of Mantid to use a newer version of Matplotlib to keep up to date with new features.

### To test:
1. Run the following tests:
   `FindPeakAutomaticTest`, `test_figureIntegration`, `test_toolbar`, `test_figuretype`, `mantidaxesTest`
2. Try to recreate the issues to see if they occur. E.g The issue with the mantidaxes was caused when a plot was plotted through the script editor, and another plot created from data was dragged onto that plot.
3.  Read my code, check for any cases I might have not thought of.

Note: getting the current colour for the `prop_cycler` has been completely removed and therefore that might be the only way of doing it even though it is not foolproof.

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
--> 

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

Despiix committed:
plotfunctionsTest.py Pass again. Getting the current colour from the axes using Prop_cycler is removed from the newer versions of Matplotlib.

Despiix committed:
Due to the ContourSet change in the 3.8 release, we can now check for contour plots by looking for the ContourSet collection in ax.collections.

Co-authored-by: Gui Pereira <gui.maciel-pereira@stfc.ac.uk>
Co-authored-by: Sam Jones <samuel.jones@stfc.ac.uk>

https://matplotlib.org/stable/api/prev_api_changes/api_changes_3.8.0.html#contourset-is-now-a-single-collection

Despiix committed:
Back to 3.8

Despiix committed:
Asserts possibly no longer necessary. They might be from a time when we had different versions of mpl on different OS.

Despiix committed:
Fix test using QuadContourSet

QuadContourSet is now a single collection instead of an artist, so I removed the call to .collections. https://matplotlib.org/3.8.4/api/prev_api_changes/api_changes_3.8.0.html#contourset-is-now-a-single-collection

Co-authored-by: Gui Pereira <gui.maciel-pereira@stfc.ac.uk>

Despiix committed:
Fixed 2 failing tests || UnboundLocalError

Turned `current_ax_colour` into a global variable to avoid the error.

Despiix committed:
Release notes

Despiix committed:
Axes3D.set_frame_on and Axes3D.get_frame_on removed

In Matplotlib V3.8 the `Axes3D.set_frame_on` has no effect on 3D axes and is being removed in favor of `Axes3D.set_axis_on` and `Axes3D.set_axis_off`.

jclarkeSTFC commented:
Would you mind adding some details about why you made the various changes, please? E.g. `Axes3D` no longer has `set_frame_on`

jclarkeSTFC reviewed (changes requested):
Looks like a good effort, I have a few questions though.

sf1919 commented:
Does this fix #37957 ?

Despiix commented on Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/FindPeakAutomaticTest.py:
```diff
@@ -485,16 +485,6 @@ def test_process_returns_the_return_value_of_find_good_peaks(self):
             self.assertTableEqual(expected_return[0][1], actual_return[0][1])
             np.testing.assert_almost_equal(expected_return[1], actual_return[1])
 
```
The tests don't fail on Windows and Ubuntu if matplotlib.pyplot is imported anymore. They are not nessesary anymore since @Pasarus said they might be from a time where we had different versions of mpl on different Operating Systems.

jclarkeSTFC replied:
So with matplotlib 3.8 `pyplot` is imported, but it isn't with 3.7, which causes this test to fail? If the other tests pass then deleting the test seems fine.

Despiix replied:
Yes, I believe that was the conclusion. All other tests pass.

Despiix commented on qt/applications/workbench/workbench/plotting/toolbar.py:
```diff
@@ -286,6 +286,8 @@ def _is_colorbar(cls, ax):
         return hasattr(ax, "_colorbar")
 
     def set_up_color_selector_toolbar_button(self, fig):
+        global current_ax_colour
```
I am sure you already know this, but I will say it just in case I am wrong.

I used a global variable because Python treats variables as local by default within a function. Without global, modifying current_ax_colour inside the function throws an UnboundLocalError since it's trying to assign it locally before initialization. Declaring it as global lets me modify the global variable. An alternative could be testing it locally by doing current_ax_colour = None instead of global current_ax_colour to declare it at the start of the function but not globally, but for some reason that doesn't work.

I am not sure about what changed that caused this error. I am assuming it's because MPL 3.8 altered how it handles color collections so the return type might be different but I am not sure.

jclarkeSTFC replied:
There are no references in our code to `current_ax_colour`, apart from in that method, and previously it was a local variable. I think we should keep that variable local instead of changing global variables, and if that doesn't work then that needs investigating.

Despiix replied:
I have just commited the fixes for this. :))

Despiix commented on qt/python/mantidqt/mantidqt/plotting/figuretype.py:
```diff
@@ -76,9 +76,7 @@ def axes_type(ax):
-        if any(isinstance(col, LineCollection) for col in ax.collections):
-            axtype = FigureType.Contour
-        elif any(isinstance(col, PathCollection) for col in ax.collections):
+        if any(isinstance(col, ContourSet) for col in ax.collections):
```
The object returned by `Contour` was a custom object holding multiple `Collections` and thus not an Artist.
In version 3.8 however, it is an Artist. So it caused an error because the return type has changed. (I forgot to take notes so I unfortunately do not remember the exact error, but I think it was an `OutOfRange` error). We can use `ContourSet` to get the contour now because it's instead a plain `Collection` which contains a single path per contour level.

jclarkeSTFC commented on qt/python/mantidqt/mantidqt/project/plotsloader.py:
```diff
@@ -306,7 +306,6 @@ def update_properties(self, ax, properties):
             ax.set_position(properties["bounds"])
         ax.set_navigate(properties["dynamic"])
         ax.axison = properties["axisOn"]
-        ax.set_frame_on(properties["frameOn"])
```
It looks to me like `set_frame_on` can be deleted for `Axes3D`, but for 2D axes that line should still be there, it's still available on Axes, see the [source](https://github.com/matplotlib/matplotlib/blob/8fb842c78f1b57b13a3d0fe63e83f3d31694f887/lib/matplotlib/axes/_base.py#L3098)

Despiix replied:
I commited a fix for this as well!

Despiix committed:
ContourSet, edgecolor changes from list to array

In the figuremanager.py the PathCollection no longer returns the paths. Instead, it has been moved to a new location where we get a QuadContourSet and change the edge colours that way instead.

In the toolbar.py the edgecolour now receives an array of identical lists instead of a single list, so the first element is picked.

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Despiix committed:
Set_Frame_On still necessary for 2D axes.

Added the set_frame_on back to the 2D axes.

jclarkeSTFC commented on qt/python/mantidqt/mantidqt/project/plotsloader.py:
```diff
@@ -305,7 +305,7 @@ def update_properties(self, ax, properties):
             ax.set_position(properties["bounds"])
         ax.set_navigate(properties["dynamic"])
-        ax.axison = properties["axisOn"]
+        ax.axisOn = properties["axisOn"]
```
```suggestion
        ax.axison = properties["axisOn"]
```
I think the property is `axison`, e.g. https://github.com/matplotlib/matplotlib/blob/a254b687df97cda8c6affa37a1dfcf213f8e6c3a/lib/matplotlib/axes/_base.py#L3495

jclarkeSTFC commented on qt/applications/workbench/workbench/plotting/toolbar.py:
```diff
@@ -305,6 +308,12 @@ def set_up_color_selector_toolbar_button(self, fig):
                 current_ax_colour = col.get_edgecolor()
                 break
+            elif isinstance(col, ContourSet):
+                current_ax_colour = col.get_edgecolor()
+                current_ax_colour = current_ax_colour[0]
```
```suggestion
                current_ax_colour = col.get_edgecolor()[0]
```

jclarkeSTFC commented on Testing/Temporary/CTestCostData.txt:
```diff
@@ -0,0 +1 @@
+---
```
I think this file should be deleted.

jclarkeSTFC commented on docs/source/release/v6.12.0/Framework/Dependencies/New_features/38077.rst:
```diff
@@ -0,0 +1 @@
+- Updated Matplotlib from version 3.7 to version 3.8.
```
```suggestion
- Updated Matplotlib from version 3.7 to version 3.8. The release notes for version 3.8 can be found `here <https://matplotlib.org/stable/users/prev_whats_new/whats_new_3.8.0.html>`_ .
```

jclarkeSTFC reviewed (changes requested):
Looks to be working fine, just a few small things before it's good to go.

Despiix committed:
Update docs/source/release/v6.12.0/Framework/Dependencies/New_features/38077.rst

Additional info on release notes

Co-authored-by: James Clarke <139879523+jclarkeSTFC@users.noreply.github.com>

Despiix committed:
Update qt/applications/workbench/workbench/plotting/toolbar.py

Removed redundunt line

Co-authored-by: James Clarke <139879523+jclarkeSTFC@users.noreply.github.com>

Despiix committed:
Minor change in axisOn  property

Co-authored-by: James Clarke <139879523+jclarkeSTFC@users.noreply.github.com>

Despiix committed:
Removed CTestCostData.txt, as it was empty

Despiix committed:
Merge remote-tracking branch 'origin/Matplotlib-3.8-Update' into Matplotlib-3.8-Update

jclarkeSTFC reviewed (approved):
Good job, thanks for making those changes.
